### DISCUSSION TITLE
How Splink is dealing with null values?

### DISCUSSION THREAD

**Discussion by Stephaniav1901** [moj-analytical-services/splink#1509] [Q&A]
when I create a blocking rule such as "l.Name = r.Name" . Is it only considering no null values? If not how do I say to only compare no null values?

**Comment by RobinL** [comment]
Yes, it only creates comparisons where the names are equal and not null

**Reply by beedrumms** [reply]
Is the behaviour the same where we have more than one attribute in a blocking rule? For example, in the case of the blocking rule `block_on("Second_Name", "City")`, would a comparison be created if either second name or city were null? Or is a comparison only created where all attributes in the blocking rule have a non-null value?

**Reply by RobinL** [reply]
See 
https://moj-analytical-services.github.io/splink/demos/examples/duckdb/deduplicate_50k_synthetic.html

````
   "blocking_rules_to_generate_predictions": [
        block_on(["first_name", "surname"]),
        block_on(["surname", "dob"]),
        block_on(["first_name", "dob"]),
        block_on(["postcode_fake", "first_name"]),
    ],

````
    
There is an implicit OR between each element in the list, so that says create all comparisons where EITHER:
    - first name and surname match OR
    - surname and sate of birth match OR 
    - etc.

**Reply by beedrumms** [reply]
Thanks so much for the response and apologies for the lack of clarity! We are trying to balance computation time against low data quality (notable proportion of missing values across all attributes and use of dependent attributes). My question is more about the treatment of nulls within each blocking rule, not between rules. Taking the example from the tutorial: 

"blocking_rules_to_generate_predictions": [
        block_on(["first_name", "surname"]),
        block_on(["surname", "dob"]),
        block_on(["first_name", "dob"]),
        block_on(["postcode_fake", "first_name"]),
    ],

If an attribute had a NULL in just the first_name field and the dob field, would it be compared at all?

**Reply by RobinL** [reply]
ah - got it.  if it has a null in `first_name`, comparisons would NOT be created for that record by the following rules:
```
block_on(["first_name", "dob"]),`
block_on(["first_name", "surname"]),
```
However, comparison should be picked up by the remaining rules e.g. `block_on(["surname", "dob"]),`

**Reply by beedrumms** [reply]
@RobinL ok noted -- thanks so much for your response :) super helpful!

**Comment by Stephaniav1901** [comment]
Nice! And Is it possible to create a blocking rule with a compare method? 
For instance, can I create a blocking rule to get the pairs where the full name has a Levenishtein distance of 10 or less?

**Reply by RobinL** [reply]
Yes here's an example:
`l.first_name = r.first_name and levenshtein(l.surname, r.surname) < 2`
But you might want to read the examples here for a bit more background:
https://github.com/moj-analytical-services/splink/discussions/1465#discussion-5438438
And here on efficiency 
https://github.com/moj-analytical-services/splink/discussions/1391
